### DISCUSSION TITLE
Next13 Middleware and rewrite to path with new request headers

### DISCUSSION THREAD

**Discussion by alecla** [vercel/next.js#45471] [Help]
I am using Next 13.1, with `app` directory and a root `middleware.ts` file. I want to redirect all requests to `/`  and also set some request headers. I noticed after setting request headers, our site will always do a full reload on navigation. 

`middleware.ts`
```js

		const requestHeaders = new Headers(request.headers);
		requestHeaders.set('x-middleware-test', 'test')
		
		if (!pathname.includes('.') && !pathname.startsWith('/api')) {
			return NextResponse.rewrite(new URL('/', request.url), {
				request: {
					headers: requestHeaders
				}
			});
		}
````

How do i set request headers and also keep Client side-navigation working for those rewrites?

**Comment by icyJoseph** [comment]
Uff I am not sure. For others reading this, the problem is that headers setting. It triggers a browser reload, and I think it makes sense, but then we lose the SPA feeling...

```ts
import { NextRequest, NextResponse } from "next/server";

export const config = {
  matcher: [
    /*
     * Match all request paths except for the ones starting with:
     * - _next/static (static files)
     * - _next/image (image optimization files)
     * - favicon.ico (favicon file)
     */
    "/((?!_next/static|_next/image|favicon.ico).*)",
  ],
};

export async function middleware(request: NextRequest) {
  console.log("Middleware");
  const requestHeaders = new Headers(request.headers);
  requestHeaders.set("x-middleware-test", "test");

  const { pathname } = request.nextUrl;

  console.log({ pathname });

  if (!pathname.includes(".") && !pathname.startsWith("/api")) {
    request.nextUrl.pathname = "/";
    const shouldUseHeaders = false;
    const options = shouldUseHeaders
      ? { request: { headers: requestHeaders } }
      : {};

    return NextResponse.rewrite(request.nextUrl, options);
  }

  return NextResponse.next();
}

```

Change the value of `shouldUseHeaders` to true, and see how the site reloads on rewrite. 

Sorry OP, not sure how to go about this one.

**Reply by alecla** [reply]
Do you know if its possible with `cookies` to avoid full reload?

**Reply by icyJoseph** [reply]
Test it :) I'll give it a go when I get home.

**Reply by icyJoseph** [reply]
Looks like doing something like:

```ts
import { NextRequest, NextResponse } from "next/server";

export const config = {
  matcher: [
    /*
     * Match all request paths except for the ones starting with:
     * - _next/static (static files)
     * - _next/image (image optimization files)
     * - favicon.ico (favicon file)
     */
    "/((?!_next/static|_next/image|favicon.ico).*)",
  ],
};

export async function middleware(request: NextRequest) {
  console.log("Middleware");

  const { pathname } = request.nextUrl;

  if (!pathname.includes(".") && !pathname.startsWith("/api")) {
    request.nextUrl.pathname = "/";

    const response = NextResponse.rewrite(request.nextUrl);

    response.cookies.set({
      name: "middleware-cookie",
      value: "foo",
      path: "/",
    });

    return response;
  }

  return NextResponse.next();
}
```

Works, without reloading the page.


https://user-images.githubusercontent.com/21013447/216094358-f0c4d35b-b143-4394-ba25-62e7300065d2.mov

**Reply by alecla** [reply]
Thanks. It seems to work great with setting response cookies like in your example. However in my case i want to set sensitive cookies that i dont want going out to the client. Thats why i wanted to use request headers.

**Reply by icyJoseph** [reply]
~~Right, you want to delete them when the request reaches the server, after middleware, right?~~

~~Do they all go through GSSP? It is possible to attach the cookie to the response going toward GSSP and delete it there, IIRC, I think I've posted that somewhere here.~~

Ah nope, you are using `appDir`, and cookies there are read only... here's what I was thinking about https://github.com/vercel/next.js/discussions/43456#discussioncomment-4254881
